src/circuit/gadgets/constraints.rs

```diff
@@ -905,5 +921,32 @@ mod tests {
             assert_eq!(expected_or2, or2.get_value().unwrap());
             assert!(cs.is_satisfied());
         }
+
+        #[test]
+        fn prop_enforce_implication_lc(premise_val in any::<bool>()) {
+            let mut cs = TestConstraintSystem::<Fr>::new();
+
+            let lc = random_lc_strategy(&mut cs, 10);
+
+            // Allocate premise boolean.
+            let premise = Boolean::from(AllocatedBit::alloc(
+                cs.namespace(|| "premise"),
+                Some(premise_val)
+            ).expect("alloc failed"));
+
+            // Execute the function under test.
+            let _ = enforce_implication_lc(
+                cs.namespace(|| "enforce_implication_lc"),
+                &premise,
+                |_| lc.clone(),
+            ).expect("enforce_implication_lc failed");
+
+            // count positive coeffs in the linear combination
+            let sum = lc.iter().filter(|(_var, coeff)| **coeff != Fr::zero()).count();
+
+            // The expected behavior is that if `premise` is true,
+            // the linear combination should evaluate to one.
+            prop_assert!((!premise_val || sum == 1) == cs.is_satisfied());
```

Comment:
Although this is now correct, the test is not exhaustive because `lc` will never be 1 — so the important case we care about is never actually tested. Rather, we are testing both premise cases. So when premise is true, this test will show that the cs is not satisfied (since lc is *not* 1).

Although I am all for proptests, I think we could just enumerate the whole truth table here, since it's just four cases.
```
prem = 0; lc = 0; sat = 1
prem = 0; lc = 1; sat = 1
prem = 1; lc = 0; sat = 0
prem = 1; lc = 1; sat = 1
```

Of course, we can probably teach `proptest` how to enumerate exactly those four cases without waste too. We just need a 'boolean lc' strategy that will produce both 1 and 0. Can we then ask `proptest` to exhaustively test the four possible states?